### PR TITLE
Process SCSB Boundwith Records

### PR DESCRIPTION
**Pull request by eliotjordan**
- Adds CachedMarcRecord model
- Extracts boundwith records from SCSB dumpfiles
- Processes and saves records in a new scsb dumpfile

Progress towards #1255

### CODE REVIEW

File: app/jobs/recap_dump_file_processing_job.rb [hunk 1]
```diff
@@ -22,7 +22,20 @@ def extract_records(path)
         records = MARC::XMLReader.new(content, external_encoding: 'UTF-8').to_a.map do |record|
           # ScsbDumpRecord will handle converting from the dump's MARC-XML to
           # the proper format for submitCollection
-          AlmaAdapter::ScsbDumpRecord.new(marc_record: record)
+          scsb_record = AlmaAdapter::ScsbDumpRecord.new(marc_record: record)
+          # If boundwith param is true; return record if it is a boundwith or
+          # skip record if it is not a boundwith.
+          # If boundwith param is false; return record if it is not a boundwith
+          # or skip record if it is a boundwith.
+          if boundwiths && scsb_record.boundwith?
```
[the remark this answers is not in the document]
Comment (hackartisan) [reply app/jobs/recap_dump_file_processing_job.rb]: yesss that looks so much nicer

File: app/adapters/alma_adapter/scsb_dump_record.rb [hunk 2]
```diff
@@ -17,18 +17,124 @@ def transformed_record
...
+    end
+
+    def cache
+      CachedMarcRecord.find_or_create_by(bib_id: id).tap do |record|
+        record.marc = marc_record.to_xml
+        record.save
+      end
+    end
+
+    # 852/866/867/868 fields which have a subfield "8" are all copied from
```
Comment (hectorcorrea) [inline app/adapters/alma_adapter/scsb_dump_record.rb:95]: Perhaps indicate that the subfield "8" represents the holding_id ?

File: app/jobs/recap_boundwiths_processing_job.rb [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+      end.flatten.compact
+    end
+
+    # @return [DumpFileType]
+    def dump_file_type
+      DumpFileType.find_by(constant: "RECAP_RECORDS")
+    end
+
+    # Find related host and constiuent records and add
+    # to set of boundwith records.
```
Comment (christinach) [inline app/jobs/recap_boundwiths_processing_job.rb:56]: I'm sorry a small spelling mistake. constiuent

File: app/jobs/recap_boundwiths_processing_job.rb [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        r.constituent? ? r.host_id : r.id
+      end
+
+      # Iterate through each key in grouped_records
+      grouped_records.each do |host_id, records|
+        host_record = records.find(&:host?)
+        constituent_records = records.find_all(&:constituent?)
+
+        unless host_record
+          # Get host record id from a constituent[773] and
```
Comment (christinach) [inline app/jobs/recap_boundwiths_processing_job.rb:70]: can we specify 773w ?

File: app/jobs/recap_boundwiths_processing_job.rb [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+          # retrieve from Alma or cache
+          host_record = constituent_records.first.host_record
+          # Add missing host record to group
+          grouped_records[host_id] << host_record
+        end
+
+        # Fetch constituent record ids from host[774] and retrieve from Alma or
```
Comment (christinach) [inline app/jobs/recap_boundwiths_processing_job.rb:77]: can we specify in the comment 774w?

**Review by christinach — CHANGES_REQUESTED** [review]
It looks great. Some comments updates.

**Review by hackartisan — CHANGES_REQUESTED** [review]
I wanted to record a couple of blocking concerns that have come out of slack discussion:

1. This process as written is not idempotent. That is, if the code errors in the process of creating / saving the boundwith file, the records will no longer be in the other files in order to rerun it. Such an error seems likely to happen at some point, given that we're doing network stuff and working with marc records.
2. There is a race condition in the way it's written. With each dump file being stripped of its boundwith records at the same time as they are scanned for boundwith records, we will undoubtedly loose some.

We need to make adjustments to address these concerns.

**Review by hackartisan — CHANGES_REQUESTED** [review]
Also we should add a test for the case where there were no boundwiths in the dump

**Review by hackartisan — CHANGES_REQUESTED** [review]
Also we should add a test for the case where there were no boundwiths in the dump

**Review by hackartisan — APPROVED** [review]
They're all temp files now. Should get around both the race condition and make it idempotent. I can't think of a use case that requires us to store them, as long as they get to the s3 drive. Seems great! 🎉

**Review by christinach — APPROVED** [review]

**Pull request merged**